### ISSUE TITLE
Enhanced Interpolations to Replace `@printf`

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaLang/julia#18703]
One way to handle the replacing `@printf` may be to enhance string interpolation. To do this, one could introduce a `to_str` function (I am not set on the name, but am using it for discussion. This is a proposal for the architecture). For example, currently `”piece1$var piece2”` parses similarly to `string(“piece1", var, “ piece2”)`. In reality, `"$var"` is the same as `"$(var)"`. My idea is to extend the interpolating `$()` to a full function call, `to_str`, and allow arguments.

For example, we could instead use `”piece1$(var,args...) piece2”` which would extend to  `string(“piece1", to_str(var,args...), “ piece2”)`. `to_str` would be a function which returns a String (or if not, `to_str` is reursively called on the output). The standard `to_str` would be

```julia
to_str{T}(x::T) = string(x)
```

which would just give the string output of the `print` method. But this let's us generalize a lot. For example, we could do

```julia
to_str{T<:AbstractFloat,}(x::T,style::Symbol;dec=2)
```

to be a function which gives back a string which which is in the chosen style. This would allow one to make `$(var,:sci,dec=3) is the result"` interpolate into the string `var` in scientific notation with 3 decimal places. 

I am not proposing the exact details for the Base functions, but I think this architecture would work really well for Base, and would be very extendable (anyone could add dispatches for their own types, so  they would work seamlessly for interpolation). This is very different from the idea of #10610, but I think this is a conservative extension to current interpolation behavior that matches well with Julia intuition and could cover most of what people need with `@printf` (with anything super special handled in packages).

**Comment by andyferris** [comment]
Does `"$x"` call `string(x)`?

Could we just allow keyword arguments:
```julia
"$(x; kwargs...)" == string(x; kwargs...)
```
And have more complex strings nest calls to "string" (one call for each $ and a final one for concatenation) or whatever.

**Comment by ChrisRackauckas** [comment]
Someone who knows more about the current implementation would have to comment. I assume that it's something similar to that, which means this wouldn't be really hard to implement. A small description of the current setup could probably make make this an easy PR.

**Comment by musm** [comment]
> Does "$x" call string(x)?

I don't think it does?

```julia
julia> x = 1
1

julia> @code_lowered("$x")
CodeInfo(:(begin
        nothing
        return (Base.dec)(x)
    end))

julia> x = 1.0
1.0

julia> @code_lowered("$x")
CodeInfo(:(begin
        nothing
        return (Core._apply)(Base.print_to_string, xs)
    end))
```
but someone more familiar with internals should comment

**Comment by vtjnash** [comment]
It does:
```which
julia> x = 1
1

julia> @which "$x"
string(x::Union{Int128, Int16, Int32, Int64, Int8}) in Base at intfuncs.jl:484

julia> x = 1.0
1.0

julia> @which "$x"
string(xs...) in Base at strings/io.jl:120

julia> expand(Main, :( "$x" ))
:((Base.string)(x))
```

**Comment by PallHaraldsson** [comment]
@ararslan Do you remember why the thumbs down? Is this just not needed in Base or some other better PR suggesting? Or as in discourse, pointed out as alternative:

https://github.com/JuliaString/StringLiterals.jl

[Should people in general explain thumps downs?]

**Comment by StefanKarpinski** [comment]
> [Should people in general explain thumps downs?]

I would generally appreciate this. An unexplained thumbs down generally comes off as kind of passive aggressive and unconstructive. Unless it's clear why you have a problem with something, it's not really helping the conversation.

**Comment by ararslan** [comment]
> Do you remember why the thumbs down?

No, but rereading this I find it odd and don't see the benefit over `@printf`.

**Comment by quinnj** [comment]
Is this still relevant/desired? I haven't seen too many requests to overload how an object is interpolated into strings, but maybe I've missed if people still need something like this. It seems that just overloading `string` for your custom type let's you hook into a wide variety of printing situations.

**Comment by simonbyrne** [comment]
A long time ago I started a Julep along these lines (https://github.com/JuliaLang/Juleps/pull/49). I'd still be in favor of this, but the main technical challenge is that it requires mucking about in the parser.

**Comment by c42f** [comment]
This isn't actually a parser error! The error arises from lowering:

```julia
julia> dump(Meta.parse("\$(1.0, digits=2)"))
Expr
  head: Symbol $
  args: Array{Any}((1,))
    1: Expr
      head: Symbol tuple
      args: Array{Any}((2,))
        1: Float64 1.0
        2: Expr
          head: Symbol =
          args: Array{Any}((2,))
            1: Symbol digits
            2: Int64 2
```

Currently syntax like `"aa$(x)bb"` lowers to `Base.string("aa", x, "bb")`, so what's involved here is figuring out a good lowering to hold the keyword arguments.

I guess I'd suggest something like `"aa$(1.0, digits=2)bb"` becoming

`Base.string("aa", Base.stringformat(1.0, digits=2), "bb")`

where `Base.stringformat()` might return a lazy wrapper which can be picked up by `Base.string` and `print`ed along with the other interpolations?

**Comment by c42f** [comment]
It's probably relatively easy to add this to the first syntax desugaring pass in lowering, somewhere around here:

https://github.com/JuliaLang/julia/blob/c4d162ea08e37dfdf958a79051aa6f6e223606b2/src/julia-syntax.scm#L2696

**Comment by simonbyrne** [comment]
Isn't that just parsing `$(1.0, digits=2)`? When you try to parse a string it throws an error:
```
julia> Meta.parse("\"\$(1.0, digits=2)\"")
ERROR: Base.Meta.ParseError("invalid interpolation syntax")
Stacktrace:
 [1] #parse#3
   @ ./meta.jl:237 [inlined]
 [2] parse(str::String; raise::Bool, depwarn::Bool)
   @ Base.Meta ./meta.jl:268
 [3] parse(str::String)
   @ Base.Meta ./meta.jl:267
 [4] top-level scope
   @ REPL[12]:1
```

**Comment by c42f** [comment]
Haha oops!  You're right of course.

**Comment by c42f** [comment]
Huh, so ah... turns out I didn't copy this restriction into JuliaSyntax.jl ... so the parser fix could be just merging https://github.com/JuliaLang/julia/pull/46372

```
julia> dump(JuliaSyntax.parsestmt(Expr, "\"\$(x, digits=1)\""))
Expr
  head: Symbol string
  args: Array{Any}((1,))
    1: Expr
      head: Symbol tuple
      args: Array{Any}((2,))
        1: Symbol x
        2: Expr
          head: Symbol =
          args: Array{Any}((2,))
            1: Symbol digits
            2: Int64 1
```

(But also I should really add this restriction back into the parser. I'm not sure how it got left out.)

**Comment by simonbyrne** [comment]
I was actually going to suggest just `:format`: the idea being it could return a wrapper that formats an object when printed or converted to a string.

**Comment by c42f** [comment]
`Expr(:format)` works for me :-)

Here's a prototype implementation in JuliaSyntax https://github.com/JuliaLang/JuliaSyntax.jl/pull/308
